File: juno/indicators/dema.py
from decimal import Decimal

from .ema import Ema


# Double Exponential Moving Average
class Dema:
    value: Decimal = Decimal("0.0")

    _ema1: Ema
    _ema2: Ema
    _t: int = 0
    _t1: int
    _t2: int

    def __init__(self, period: int) -> None:
        self._ema1 = Ema(period)
        self._ema2 = Ema(period)
        self._t1 = period
        self._t2 = period * 2 - 1

    @property
    def maturity(self) -> int:
        return self._t2

    @property
    def mature(self) -> bool:
        return self._t >= self._t2

    def update(self, price: Decimal) -> Decimal:
        self._t = min(self._t + 1, self._t2)

        self._ema1.update(price)

        if self._t <= self._t1:
            self._ema2.update(price)

        if self._t >= self._t1:
            self._ema2.update(self._ema1.value)
            if self._t >= self._t2:
                self.value = self._ema1.value * Decimal("2.0") - self._ema2.value

        return self.value

File: juno_rs/indicators/dema.rs
use super::{ema::Ema, MA};
use std::cmp::min;

pub struct Dema {
    pub value: f64,
    ema1: Ema,
    ema2: Ema,
    t: u32,
    t1: u32,
    t2: u32,
}

impl Dema {
    pub fn new(period: u32) -> Self {
        let t1 = period - 1;
        Self {
            value: 0.0,
            ema1: Ema::new(period),
            ema2: Ema::new(period),
            t: 0,
            t1,
            t2: t1 * 2,
        }
    }

    pub fn maturity(&self) -> u32 {
        self.t2
    }

    pub fn update(&mut self, price: f64) {
        self.ema1.update(price);

        if self.t <= self.t1 {
            self.ema2.update(price);
        }

        if self.t >= self.t1 {
            self.ema2.update(self.ema1.value);
            if self.t == self.t2 {
                self.value = self.ema1.value * 2.0 - self.ema2.value;
            }
        }

        self.t = min(self.t + 1, self.t2);
    }
}

impl MA for Dema {
    fn update(&mut self, price: f64) {
        self.update(price)
    }

    fn value(&self) -> f64 {
        self.value
    }

    fn maturity(&self) -> u32 {
        self.t2
    }
}
